### PR TITLE
Add retries to CLI test framework for network errors

### PR DESCRIPTION
**Pull request by dubee**
Allows the CLI test framework to perform retries when a network error occurs.

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [x] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

File: tests/src/test/scala/common/TestUtils.java [hunk 1]
```diff
@@ -54,11 +54,12 @@
...
-    public static final int ERROR_EXIT      = 1;
-    public static final int MISUSE_EXIT     = 2;
-    public static final int DONTCARE_EXIT   = -1;       // any value is ok
-    public static final int ANY_ERROR_EXIT  = -2;       // any non-zero value is ok
+    public static final int SUCCESS_EXIT        = 0;
+    public static final int ERROR_EXIT          = 1;
+    public static final int MISUSE_EXIT         = 2;
+    public static final int DONTCARE_EXIT       = -1;       // any value is ok
+    public static final int ANY_ERROR_EXIT      = -2;       // any non-zero value is ok
+    public static final int NETWORK_ERROR_EXIT  = 3;
```
Comment (markusthoemmes) [inline tests/src/test/scala/common/TestUtils.java]: Which errors are part of this? Does the CLI return a "3" on network issues as a convention?

**Comment by markusthoemmes** [comment]
General existential question: I'm unclear if this changes any semantics in our tests. The main issue I could see (and I'm not sure if its really a problem) is that our requests are not necessarily idempotent. That is: A connection error might happen at any stage of the connection so you'll never know if it's safe to retry.

This might still make things better but it could certainly harm debuggability in tests. It might make sense to add a logline stating it's retrying because of a network issue to be able to tell later if an inconsistency was caused by that. Thoughts?

File: tests/src/test/scala/common/TestUtils.java (hunk 1 above)
Comment (mdeuser) [reply tests/src/test/scala/common/TestUtils.java]: yes, when detected.
https://github.com/apache/incubator-openwhisk-client-go/blob/df32dca4aecca4ec30add11ce89eac72928f95f4/whisk/wskerror.go#L22

File: tests/src/test/scala/common/Wsk.scala [hunk 2]
```diff
@@ -1036,13 +1036,26 @@ trait RunWskCmd extends BaseRunWsk {
...
+          val rr = TestUtils.runCmd(
+            DONTCARE_EXIT,
+            workingDir,
+            TestUtils.logger,
+            sys.env ++ env,
+            stdinFile.getOrElse(null),
+            args ++ params: _*)
+
+          if (expectedExitCode != NETWORK_ERROR_EXIT && rr.exitCode == NETWORK_ERROR_EXIT) {
+            println("A network error occurred.")
```
Comment (mdeuser) [inline tests/src/test/scala/common/Wsk.scala]: maybe add a bit more context to this output line... i.e. that the cli will be retried.. maybe the retry count if available..

File: tests/src/test/scala/common/Wsk.scala [hunk 3]
```diff
@@ -1036,13 +1036,28 @@ trait RunWskCmd extends BaseRunWsk {
...
+          val rr = TestUtils.runCmd(
+            DONTCARE_EXIT,
+            workingDir,
+            TestUtils.logger,
+            sys.env ++ env,
+            stdinFile.getOrElse(null),
+            args ++ params: _*)
+
+          if (expectedExitCode != NETWORK_ERROR_EXIT) {
+            withClue(hideStr(reportFailure(args ++ params, expectedExitCode, rr).toString(), hideFromOutput)) {
```
Comment (mdeuser) [inline tests/src/test/scala/common/Wsk.scala]: is the `withClue` line still needed now that the `retry` is displaying the failure message?

Comment (dubee) [reply tests/src/test/scala/common/Wsk.scala]: Yes, `withClue` will print the stack trace if the final retry fails.

Comment (mdeuser) [reply tests/src/test/scala/common/Wsk.scala]: will each retry will have a `withClue` stack trace message, followed by a final `withClue` message (line 1062) if all retries fail?

Comment (dubee) [reply tests/src/test/scala/common/Wsk.scala]: Just a final retry that fails will have the stack trace.

**Comment by dubee** [comment]
@mdeuser, anything else on this one?

**Comment by dubee** [comment]
PG4 1807 ⏳

**Comment by mdeuser** [comment]
pg :large_blue_circle: :+1:

File: tests/src/test/scala/common/Wsk.scala (hunk 3 above)
Comment (rabbah) [reply tests/src/test/scala/common/Wsk.scala]: This change in my opinion is wrong and should be backed out.

The retry method used here requires the function under retry to throw an exception to be retried, and use of `retry` here makes it impossible to exit the retry loop subject to the expected exit code unless that code is `NETWORK_ERROR_EXIT`.

I suggest backing this out and then fixing it properly, with a test for the `cli` command itself.

FYI @houshengbo

Comment (rabbah) [reply tests/src/test/scala/common/Wsk.scala]: I'm addressed this in #3776.

**Pull request merged**